src/C-interface/bml_adjungate_triangle.c

```diff
@@ -16,7 +16,7 @@
  */
 void
 bml_adjungate_triangle(
-    bml_matrix_t * const A,
+    struct bml_matrix_t *const A,
```

Comment:
@jmohdyusof I think the reason for the additional space is that indent doesn't see bml_matrix_t as a type. We could add the `struct` keyword so that `indent` "gets it". What do you think?

Reply:
Hmm, this solution doesn't work since the `typedef void bml_matrix_t`. Too bad :frowning_face:

Reply:
Thinking out loud: what is the benefit of declaring the pointer const if all the data in the struct is being modified (potentially)? Is there anything the compiler can do to improve performance when only the pointer is const?

Reply:
The address of `a` in `void *a = malloc(); foo(a);` is passed by value to the function `void foo(void *a);`. The function `foo` can not communicate any change of that address back to the caller, and in that sense the definition `void foo(void *a)` and `void foo(void *const a)` are identical. The only difference with adding a `const` is that the compiler can throw an error if that address is changed inside the function. I don't know if that means it could optimize something, but since it knows now that the value of that pointer can never change it might for instance hold that pointer value in a register. But that's just a guess. I don't really understand compilers :smile:

Reply:
I don't really understand them either: when we apply const to the data itself then I can see the value of allowing the compiler to know there is no aliasing, putting the array in read-only storage on GPU etc. Just less clear of the benefits for the pointer. Oh well.

Reply:
To me it looks like the C compiler is ignoring these 'const'. In particular, it allows changes inside a function when declaring a const...

Reply:
gcc does not ignore the `const` @jeanlucf22 :

```
void foo(void *const p)
{
  p = 0;
}
```
does not compile.

Reply:
@jmohdyusof From what I understand, the original problem you ran into on GPUs was a  `const` on a matrix in `bml_add` that was supposed to receive the result, and not a missing `const` on a pointer, right?

The more I think about it, the less I see the value in adding a `const` to the pointer itself. As you pointed out, it's hard to imagine what kind of optimizations the compiler might be able to perform with `const` that it can't do without it. And the additional compiler checks that ensure that we don't accidentally modify the pointer itself hardly seem to be worth the effort of adding a `const` all over the place either.

What does everyone think about just making sure that the `const` on the referenced matrix is correct and to not add `const` on the pointer itself? @cnegre , @jeanlucf22 , @suemni ?

Reply:
> gcc does not ignore the `const` @jeanlucf22 :
> 
> ```
> void foo(void *const p)
> {
>   p = 0;
> }
> ```
> 
> does not compile.

Indeed. But code like this works right now:

void TYPED_FUNC(
    bml_copy_ellpack) (
    const bml_matrix_ellpack_t * A,
    const bml_matrix_ellpack_t * B)
{
    REAL_T *B_value = B->value;
...
        memcpy(&B_value[ROWMAJOR(i, 0, N, M)], &A_value[ROWMAJOR(i, 0, N, M)],
...

So we can modify an element of a const struct*. (I believe you could not do that in C++)
Then I don't see the point of using const at all...

Reply:
@nicolasbock I don't recall which function it was, but yes, the main point was that *some* compilers (correctly) complained that we were modifying parts of a struct that had been declared const. 
I do not think I have ever seen a compiler complain about a missing const (I am not sure what criterion a compiler could use to determine that a const was required, apart from a discrepancy between the code and a previous declaration in a .h file?)

Reply:
@jeanlucf22 Thanks for pointing that out, I think you are right. The issue seems to be that the `const` only declares the elements of the `struct` to be constants. That means in your example that the pointer `B->value` is `const` and we wouldn't be able to `B->value = 0;`. But the compiler ignores the memory the pointer points to and we can modify that. Not quite what we had hoped :disappointed: .

I am not very familiar with C++ to make an educated statement to whether this would be possible in C++ or not, but you might very well be right.

I found this [blog post](https://www.airs.com/blog/archives/428) which sums it up nicely I think. Basically `const` applied to a pointer is not always enforced by the compiler. One can type cast the `const` away or one can use a second pointer to point to the same memory and modify it.

When I wrote the code and used `const` I wasn't quite aware of those shortcomings. My intention was to have the compiler help a little to ensure that we don't modify anything that shouldn't be modified. Kind of what one expects when using a word such as `const` :smile:. But given the shortcomings, should we just simplify things and remove all `const`?

Reply:
@jmohdyusof Interesting observations. Sounds like at least some compilers are a little careful when checking `const` :smile: Removing the `const` should be more compatible and certainly not wrong.

Reply:
I think all the compilers (hopefully) will catch the inconsistent declaration issue, but as was pointed out, const pointers are kind of special. Removing the const is fine with me, we can leave them where the struct data is not modified if we want?

Reply:
Perfect! I would prefer to make things consistent and remove all `const`. Also, since we can't quite trust the compiler to catch us modifying struct data, a `const` might give us a wrong sense of security.

Reply:
https://stackoverflow.com/questions/28062095/pass-a-two-dimensional-array-to-a-function-of-constant-parameter/28062262#28062262

Reply:
I don't know, I have read this now three times and still don't quite get it. But it's clear that using `const` is pretty complicated and error prone. Better get rid of it :smile: